[c10d] Fix extra CUDA context created by barrier

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149144

Fixes #149119.

In ProcessGroup.hpp, we create a dummy tensor for dispatching. This
requires a correct device index. This PR uses `device_id` given by user
when calling `init_process_group`.

This PR also uses `torch._C._get_accelerator()` to determine the device
type.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[c10d] Fix extra CUDA context created by barrier

Fixes #149119.

In ProcessGroup.hpp, we create a dummy tensor for dispatching. This
requires a correct device index. This PR uses `device_id` given by user
when calling `init_process_group`.

This PR also uses `torch._C._get_accelerator()` to determine the device
type.

[ghstack-poisoned]

d4l3k commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4654,30 +4654,43 @@ def barrier(
+    elif group.bound_device_id is not None:
+        # Use device id from `init_process_group(device_id=...)`
+        opts.device = group.bound_device_id
+    elif device.type == "cpu" or get_backend(group) == Backend.GLOO:
```
Is there a way to avoid depending on specific backend names/types? This makes it hard to add new ones that are compatible with core PT -- I've been trying to clean these up for torchft

kwen2501 replied:
Yeah, I hope there is a way. The specific code is for a case where the user is on a GPU machine but only want to use CPU to do some stuff...

d4l3k commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4654,30 +4654,43 @@ def barrier(
             the default process group will be used.
         async_op (bool, optional): Whether this op should be an async op
-        device_ids ([int], optional): List of device/GPU ids.
+        device_ids ([int], optional): List of device/GPU ids. Only one id is expected.
```
Can we change this to

```
Only the first ID is used.
```

kwen2501 replied:
I do mean only one is expected, because now we are expecting one device per thread. Some of the API signatures came from the old days.

XilunWu reviewed (approved):
LGTM, thx for the fix!

d4l3k commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4654,30 +4654,43 @@ def barrier(
+        # Use device id from `init_process_group(device_id=...)`
+        opts.device = group.bound_device_id
+    elif device.type == "cpu" or get_backend(group) == Backend.GLOO:
+        opts.device = torch.device("cpu")
```
Will Gloo fail if it's not a CPU device?

kwen2501 committed:
Update on "[c10d] Fix extra CUDA context created by barrier"


Fixes #149119.

In ProcessGroup.hpp, we create a dummy tensor for dispatching. This
requires a correct device index. This PR uses `device_id` given by user
when calling `init_process_group`.

This PR also uses `torch._C._get_accelerator()` to determine the device
type.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
Failure seems to be an issue of CI instance and unrelated.
@pytorchbot merge -i

cyyever commented:
@pytorchbot merge -f "Unrelated failures"

kwen2501 committed:
Update on "[c10d] Fix extra CUDA context created by barrier"


Fixes #149119.

In ProcessGroup.hpp, we create a dummy tensor for dispatching. This
requires a correct device index. This PR uses `device_id` given by user
when calling `init_process_group`.

This PR also uses `torch._C._get_accelerator()` to determine the device
type.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Internal failure looks legit' -c ghfirst

```
E           Traceback of where the remote function was issued on controller (most recent call last):
E             <not related to a specific invocation>
E           Traceback of where the remote function failed on worker (most recent call last):
E             File "<unknown>", line None, in remote function failed: Traceback (most recent call last):
E             File "/re_cwd/buck-out/v2/gen/fbcode/d6a07959eaa8ed59/monarch/python/tests/__test_remote_functions__/test_remote_functions#link-tree/monarch/worker/_testing_function.py", line 101, in barrier
E               dist.barrier(group=group, async_op=False, device_ids=device_ids)
E             File "/re_cwd/buck-out/v2/gen/fbcode/d6a07959eaa8ed59/monarch/python/tests/__test_remote_functions__/test_remote_functions#link-tree/torch/distributed/c10d_logger.py", line 81, in wrapper
E               return func(*args, **kwargs)
E             File "/re_cwd/buck-out/v2/gen/fbcode/d6a07959eaa8ed59/monarch/python/tests/__test_remote_functions__/test_remote_functions#link-tree/torch/distributed/distributed_c10d.py", line 4770, in barrier
E               work = group.barrier(opts=opts)
E           <class 'RuntimeError'>: CUDA error: invalid device ordinal
E           CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
E           For debugging consider passing CUDA_LAUNCH_BLOCKING=1
E           Device-side assertion tracking was not enabled by user.
E           RuntimeError: remote function failed: Traceback (most recent call last):
E             File "/re_cwd/buck-out/v2/gen/fbcode/d6a07959eaa8ed59/monarch/python/tests/__test_remote_functions__/test_remote_functions#link-tree/monarch/worker/_testing_function.py", line 101, in barrier
E               dist.barrier(group=group, async_op=False, device_ids=device_ids)
E             File "/re_cwd/buck-out/v2/gen/fbcode/d6a07959eaa8ed59/monarch/python/tests/__test_remote_functions__/test_remote_functions#link-tree/torch/distributed/c10d_logger.py", line 81, in wrapper
E               return func(*args, **kwargs)
E             File "/re_cwd/buck-out/v2/gen/fbcode/d6a07959eaa8ed59/monarch/python/tests/__test_remote_functions__/test_remote_functions#link-tree/torch/distributed/distributed_c10d.py", line 4770, in barrier
E               work = group.barrier(opts=opts)
E           <class 'RuntimeError'>: CUDA error: invalid device ordinal
E           CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
E           For debugging consider passing CUDA_LAUNCH_BLOCKING=1
E           Device-side assertion tracking was not enabled by user.
```

kwen2501 commented:
@pytorchbot merge -f "Internal test was wrong; OSS version of barrier tests passed"
